### PR TITLE
ci : Add github action workflow for okd 4.14.0 cluster with crc

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Add workflow for OKD 4.14.0 for running Fabric8 Kubernetes Client E2E tests.

+ We create crc cluster with a custom okd preset, that creates a cluster based on [OpenShift Kubernetes Distribution](https://okd.io/)
+ CRC version is bound to OKD, for future upgrades we would need to bump CRC version.
+ I managed to get it working with CRC 2.32.0, I couldn't get it working with latest version of CRC, I've created https://github.com/crc-org/crc/issues/4382 on crc repository
+ Update existing tests for running on OpenShift Kubernetes Distribution 4.x:
   - add support for handing `networking.k8s.io/v1` Ingress in ServiceToURLIT, replace deprecated methods
   - Add an exception to PluralizeIT for `ResourceAccessReview` resource, OpenShift Aggregated Discovery API is returning `singularName` for this resource as `localresourceaccessreview`, that does not match test expectations
   - Relax assertion check in ApiVersionsIT to not match exact hostname for server address

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: .github/workflows/e2e-tests.yml [hunk 1]
```diff
@@ -154,3 +154,72 @@ jobs:
...
+    runs-on: ubuntu-latest
+    if: github.repository == 'fabric8io/kubernetes-client'
+    env:
+      SHELL: /bin/bash
+    strategy:
+      fail-fast: false
+      matrix:
+        # OKD version is bound to crc version
+        okd: [v4.14.0]
+        crc: [2.32.0]
```
Comment (manusa) [inline .github/workflows/e2e-tests.yml:169]: Please add a comment as to why these versions aren't the latest.
Also create an issue for further investigation and future updates and put the link in the comment.

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -154,3 +154,72 @@ jobs:
...
+          sudo rm -rf "/usr/local/share/boost"
+          sudo rm -rf /opt/hostedtoolcache/CodeQL
+
+          sudo docker image prune --all --force
+
+          sudo swapoff -a
+          sudo rm -f /mnt/swapfile
+          sudo adduser $USER libvirt
+          sudo adduser $USER kvm
+          sudo usermod -a -G libvirt $USER
```
Comment (manusa) [inline .github/workflows/e2e-tests.yml:202]: Have you tried not to include this extra-step?

**Comment by manusa** [comment]
Since this takes ~30 minutes to run, it might be better to set it in a different workflow file and probably run only nightly.
We can consider making this per-pullrequest triggered once it gets quicker or it becomes more critical.

File: .github/workflows/e2e-tests.yml (hunk 2 above)
Comment (rohanKanojia) [reply .github/workflows/e2e-tests.yml:202]: When I try to remove this step in my [test repository github action](https://github.com/rohankanojia-forks/crc-okd-4-15-github-action-poc/actions/runs/11178021397/job/31074742305), I see this error:
```
write /home/runner/.crc/cache/tmp-extract/crc_okd_libvirt_4.14.0-0.okd-scos-2024-01-10-151818_amd64/crc.qcow2: no space left on device
Error: Process completed with exit code 1.
```

File: .github/workflows/e2e-tests.yml (hunk 1 above)
Comment (rohanKanojia) [reply .github/workflows/e2e-tests.yml]: I have created this issue https://github.com/fabric8io/kubernetes-client/issues/6415. It's also referenced in TODO comment.

File: .github/workflows/e2e-crc-okd-tests.yml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        run: |
+          crc config set preset okd
+          crc config set network-mode user
+      - name: Setup the crc
+        run: |
+          sudo -su $USER crc setup
+      - name: Start the crc
+        run: sudo -su $USER crc start
+      - name: Install Kubernetes Client
+        run: ./mvnw ${MAVEN_ARGS} clean install -DskipTests -Djacoco.skip=true
```
Comment (manusa) [inline .github/workflows/e2e-crc-okd-tests.yml:106]: I`make quickly` should be used here instead.

Especially if we intend to run this for each PR

File: .github/workflows/e2e-crc-okd-tests.yml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+env:
+  IT_MODULE: kubernetes-itests
+  MAVEN_ARGS: -B -C -V -ntp -Dhttp.keepAlive=false -e
+  SHELL: /bin/bash
+
+jobs:
+  openshift-kubernetes-distribution:
+    name: CRC ${{ matrix.crc }} / OKD ${{ matrix.okd }}
+    runs-on: ubuntu-24.04
```
Comment (manusa) [inline .github/workflows/e2e-crc-okd-tests.yml:41]: Won't this run faster in macos?

Comment (rohanKanojia) [reply .github/workflows/e2e-crc-okd-tests.yml:41]: I checked with crc team about this. They had tried it in past but it didn't work due to limited resources available on macOS runner. I'll check it again and see if it's possible.

**Comment by manusa** [comment]
It is still failing though

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**